src/transforms/transaction.rs

```diff
@@ -0,0 +1,812 @@
+                output.push(Event::from(t.flush()));
+            }
+        }
+    }
+
+    fn flush_all_into(&mut self, output: &mut Vec<Event>) {
+        self.transaction_merge_states
+            .drain()
+            .for_each(|(_, s)| output.push(Event::from(s.flush())));
+    }
+}
+
+impl Transform for Transaction {
+    // Only used in tests
+    fn transform(&mut self, event: Event) -> Option<Event> {
+        let mut output = Vec::new();
+        self.transform_into(&mut output, event);
+        output.pop()
+    }
+
+    fn transform_into(&mut self, output: &mut Vec<Event>, event: Event) {
+        let ends_here = self
+            .ends_when
+            .as_ref()
+            .map(|c| c.check(&event))
+            .unwrap_or(false);
+
+        let event = event.into_log();
+        let discriminant = Discriminant::from_log_event(&event, &self.identifier_fields);
+
+        if ends_here {
+            output.push(Event::from(
+                if let Some(mut state) = self.transaction_merge_states.remove(&discriminant) {
+                    state.add_event(event, &self.merge_strategies);
+                    state.flush()
+                } else {
+                    TransactionState::new(event, &self.merge_strategies).flush()
+                },
+            ));
+        } else {
+            match self.transaction_merge_states.entry(discriminant) {
+                hash_map::Entry::Vacant(entry) => {
+                    entry.insert(TransactionState::new(event, &self.merge_strategies));
+                }
+                hash_map::Entry::Occupied(mut entry) => {
+                    entry.get_mut().add_event(event, &self.merge_strategies);
+                }
+            }
+        }
+
+        self.flush_into(output);
+    }
+
+    fn transform_stream(
+        self: Box<Self>,
+        input_rx: Box<dyn Stream01<Item = Event, Error = ()> + Send>,
+    ) -> Box<dyn Stream01<Item = Event, Error = ()> + Send>
+    where
+        Self: 'static,
+    {
+        let mut me = self;
+
+        let poll_period = me.flush_period.clone();
+
+        let mut flush_stream = tokio::time::interval(poll_period);
+        let mut input_stream = Compat01As03::new(input_rx);
+
+        let stream = stream! {
+          loop {
+            let mut output = Vec::new();
+            let done = tokio::select! {
+                _ = flush_stream.next() => {
+                  me.flush_into(&mut output);
+                  false
+                }
+                maybe_event = input_stream.next() => {
+                  match maybe_event {
+                    None => {
+                      me.flush_all_into(&mut output);
+                      true
+                    }
+                    Some(Ok(event)) => {
+                      me.transform_into(&mut output, event);
+                      false
+                    }
+                    Some(Err(())) => unreachable!(),
```

Comment:
Are you sure? Can you document why?

Reply:
Changing to a panic with message. This is just reading from a channel, so the unit error should be impossible. It will go away with new streams that don't require an error type.